### PR TITLE
Manage notification templates

### LINKED ISSUE #127: As an admin I want to manage email templates so I can control the messages that users receive
Some of possible templates:
- [ ] Registration success 
- [ ] Account blocked/unblocked
- [ ] Event blocked/unblocked
- [ ] New event in favorites
- [ ] Event you're gonna visit is modified
- [ ] etc...

### PR DESCRIPTION
**Pull request by khvatskyi**
dev
## GitHub Project

* [Main GitHub Project ticket](https://github.com/EventsExpress/EventsExpress/projects/2)

## Code reviewers

- [ ] @iuriikhrystiuk 

### Second Level Review

- [ ] @sand0 

## Summary of issue

Notification templates are hard-coded. There is a need to create the ability to change the body of the template.

## Summary of change

At the DAL level, an entity has been created to store notification templates. Organized default initialization.
At the BLL level, a notification service was created and notification handlers were reconstructed.
At the API level, a NotificationController has been created, and also MapperProfile has been added.
On the front end, in a nutshell, created functionality that allows the administrator to manage notification templates.

## Testing approach

Unit tests have been written for each module.

## CHECK LIST
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

### CODE REVIEW

File: EventsExpress.Core/DTOs/NotificationTemplateDTO.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+﻿// unset
```
Comment (iuriikhrystiuk) [inline EventsExpress.Core/DTOs/NotificationTemplateDTO.cs]: Consider not having such comments. Remove it please.

File: EventsExpress.Core/IServices/INotificationTemplateService.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using EventsExpress.Core.DTOs;
+using EventsExpress.Db.Entities;
+
+namespace EventsExpress.Core.IServices
+{
+    public interface INotificationTemplateService
+    {
+        public Task<NotificationTemplateDTO> AddAsync(NotificationTemplateDTO notificationTemplateDto);
```
Comment (iuriikhrystiuk) [inline EventsExpress.Core/IServices/INotificationTemplateService.cs]: You do not add any templates manually. Remove this from interface and implementation.

File: EventsExpress.Core/IServices/INotificationTemplateService.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+        public Task<IEnumerable<NotificationTemplateDTO>> GetAllAsync();
+
+        public Task<IEnumerable<NotificationTemplateDTO>> GetAsync(int page, int pageSize);
+
+        public Task<NotificationTemplateDTO> GetByIdAsync(Guid id);
```
Comment (iuriikhrystiuk) [inline EventsExpress.Core/IServices/INotificationTemplateService.cs]: The templates should not have autogenerated id. We control the templates and we use them from the backend, so we need to know the identifier. The Id in this case should an enum value.

File: EventsExpress.Core/IServices/INotificationTemplateService.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+        public Task<NotificationTemplateDTO> GetByTitleAsync(string title);
```
Comment (iuriikhrystiuk) [inline EventsExpress.Core/IServices/INotificationTemplateService.cs]: Then you don't need to search by title anymore.

File: EventsExpress.Core/Services/NotificationTemplateService.cs [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+            NotificationTemplate notificationTemplate = Mapper.Map<NotificationTemplate>(notificationTemplateDto);
+            Insert(notificationTemplate);
+            await Context.SaveChangesAsync();
+
+            notificationTemplateDto = await GetByTitleAsync(notificationTemplate.Title);
+
+            return notificationTemplateDto;
+        }
+
+        public async Task DeleteByIdAsync(Guid id)
```
Comment (iuriikhrystiuk) [inline EventsExpress.Core/Services/NotificationTemplateService.cs]: You are not deleting the template either. Remove this method please.

File: EventsExpress.Core/Services/NotificationTemplateService.cs [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        {
+            var template = Mapper.Map<NotificationTemplate>(GetByIdAsync(id));
+            Delete(template);
+            await Context.SaveChangesAsync();
+        }
+
+        public async Task<IEnumerable<NotificationTemplateDTO>> GetAllAsync()
```
Comment (iuriikhrystiuk) [inline EventsExpress.Core/Services/NotificationTemplateService.cs]: Why do you need the GetAll method?

File: EventsExpress.Db/EventsExpress.Db.csproj [hunk 7]
```diff
@@ -17,6 +17,9 @@
       <IncludeAssets>runtime; build; native; contentfiles; analyzers</IncludeAssets>
```
Comment (iuriikhrystiuk) [inline EventsExpress.Db/EventsExpress.Db.csproj]: No need to change the csproj file. Undo it please.

File: EventsExpress.Db/Migrations/20210413112849_AddNotificationTemplate.cs [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    public partial class AddNotificationTemplate : Migration
+    {
+        protected override void Up(MigrationBuilder migrationBuilder)
+        {
+            migrationBuilder.CreateTable(
+                name: "NotificationTemplates",
+                columns: table => new
+                {
+                    Id = table.Column<Guid>(nullable: false),
+                    Title = table.Column<string>(nullable: true),
```
Comment (iuriikhrystiuk) [inline EventsExpress.Db/Migrations/20210413112849_AddNotificationTemplate.cs]: All of the string columns are required. Please make the DB so as well.

File: EventsExpress/ClientApp/src/components/app/app.js [hunk 9]
```diff
@@ -24,6 +24,8 @@ import AddEventWrapper from '../../containers/add-event';
 import Admin from '../admin';
 import DraftEditWrapper from '../../containers/draft-edit-wrapper';
 import EventDraftListWrapper from '../../containers/event-draft-list'
+import NotificationTemplateWrapper from "../../containers/notification-template/notification-template";
```
Comment (iuriikhrystiuk) [inline EventsExpress/ClientApp/src/components/app/app.js]: These imports are not used anywhere. Consider undoing this file.

File: EventsExpress/ClientApp/src/components/notification_template/notification-template-list.js [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+import { connect } from 'react-redux';
+import NotificationTemplateItem from './notification-template-item';
+
+export default class NotificationTemplates extends Component {
+    
+    renderTemplates = (arr) => arr.map(
+        template => <NotificationTemplateItem key={template.id} template={template} />
+    );
+    
+    changeMessage = () => {
```
Comment (iuriikhrystiuk) [inline EventsExpress/ClientApp/src/components/notification_template/notification-template-list.js]: Remove unused methods please.

File: EventsExpress/ClientApp/src/services/NotificationTemplateService.js [hunk 11]
```diff
@@ -0,0 +1,24 @@
+﻿import EventsExpressService from "./EventsExpressService";
+
+const baseService = new EventsExpressService();
+
+export default class NotificationTemplateService {
+
+    addAsync = async (data) => {
```
Comment (iuriikhrystiuk) [inline EventsExpress/ClientApp/src/services/NotificationTemplateService.js]: Remove the add method please.

File: EventsExpress/ClientApp/src/services/NotificationTemplateService.js [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+        return !res.ok
+            ? { error: await res.text() }
+            : res;
+    }
+
+    getAll = (pageNumber = 1, pageSize = 10) => baseService.getResourceNew(`NotificationTemplate/All?page=${pageNumber}&pageSize=${pageSize}`);
+    
+    getByIdAsync = id => baseService.getResourceNew(`NotificationTemplate/${id}/Get`);
+
+    updateAsync = async data => {
```
Comment (iuriikhrystiuk) [inline EventsExpress/ClientApp/src/services/NotificationTemplateService.js]: Consider using the new approach with the service without the ternary operators in the return statement.

File: EventsExpress/Controllers/NotificationTemplateController.cs [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+using System.Threading.Tasks;
+using EventsExpress.Core.DTOs;
+using EventsExpress.Core.IServices;
+using Microsoft.AspNetCore.Mvc;
+
+namespace EventsExpress.Controllers
+{
+    [Route("api/[controller]")]
+    [ApiController]
+    public class NotificationTemplateController : Controller
```
Comment (iuriikhrystiuk) [inline EventsExpress/Controllers/NotificationTemplateController.cs:16]: What about authentication?

File: EventsExpress/Controllers/NotificationTemplateController.cs [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+
+        public NotificationTemplateController(INotificationTemplateService notificationTemplateService)
+        {
+            _notificationTemplateService = notificationTemplateService;
+        }
+
+        [HttpPost("Add")]
+        public async Task<ActionResult> Add(NotificationTemplateDTO notificationTemplateDto)
+        {
+            if (!ModelState.IsValid)
```
Comment (iuriikhrystiuk) [inline EventsExpress/Controllers/NotificationTemplateController.cs]: We don't do the `ModelState` validation in controllers. It is automatically done in the binder.

**Review by iuriikhrystiuk — COMMENTED** [review]
You don't have the validation either in your forms or in your API. Please add it as well.

File: EventsExpress/Controllers/NotificationTemplateController.cs [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+        private readonly IMapper _mapper;
+
+        public NotificationTemplateController(INotificationTemplateService notificationTemplateService, IMapper mapper)
+        {
+            _notificationTemplateService = notificationTemplateService;
+            _mapper = mapper;
+        }
+
+        [HttpGet("All")]
+        public async Task<ActionResult<IEnumerable<NotificationTemplateDTO>>> GetAll([FromQuery] int page, [FromQuery] int pageSize)
```
Comment (iuriikhrystiuk) [inline EventsExpress/Controllers/NotificationTemplateController.cs]: Just noticed, that you have paging set up in controller, but there is no page control on UI. Let's remove the paging functionality from controller and router altogether.

**Review by iuriikhrystiuk — DISMISSED** [review]

File: EventsExpress.Test/EventsExpress.Test.csproj [hunk 16]
```diff
@@ -6,6 +6,10 @@
     <IsPackable>false</IsPackable>
   </PropertyGroup>
 
+  <ItemGroup>
```
Comment (iuriikhrystiuk) [inline EventsExpress.Test/EventsExpress.Test.csproj]: Undo changes to this file.

File: EventsExpress/ClientApp/package.json [hunk 17]
```diff
@@ -45,7 +45,7 @@
         "react-moment": "0.9.2",
         "react-pager": "1.3.3",
         "react-paginating": "1.3.0-hook.2",
-        "react-phone-number-input": "3.1.17",
+        "react-phone-number-input": "^3.1.17",
```
Comment (iuriikhrystiuk) [inline EventsExpress/ClientApp/package.json]: Undo this file.

Comment (khvatskyi) [reply EventsExpress/ClientApp/package.json]: Done

File: EventsExpress.Test/EventsExpress.Test.csproj (hunk 16 above)
Comment (khvatskyi) [reply EventsExpress.Test/EventsExpress.Test.csproj]: Done

**Review by iuriikhrystiuk — APPROVED** [review]

**Pull request merged**